# 2024-05-07 (수) 진행상황
## 프론트엔드 & AI
- AI를 적용해야했다. 그래서 일단 AI 담당자(최민석)에게 모델을 업로드해달라고 요청하고, 해당 모델을 내 PC에 받아서 사용하기로 했다.
- Python환경에서 돌아가야하기에 관련된 모듈을 모두 다운받고 서버를 실행시켰고, 127.0.0.1:5000 으로 서버가 open 되었다.
- 우리의 최종 목표는 앱 프론트엔드에서 플라스크로 열린 서버에 사진과 함께 axios post 요청을 하면, 해당 결과를 AI서버에게 받아서 프론트엔드에게 알려주는것이다.
- 우선 프론트엔드에서 useFaceDetection이라는 훅을 만들고, FormData에 담아서 post요청을 해보았다.
- 자꾸 안된다?? 생각해보니 포트포워딩을 해줘야하기에 8080말고도 5000도 포트포워딩을 새롭게 추가했다.
- 근데 추가해도 안된다!! 아예 python서버에 로그 자체가 안찍힘! AI 모델을 분명 postman에서는 확인했는데 안되는게 이상했다..
- https://yuip.org/en/port-check 라는 사이트에서 나의 ip에서 포트가 열려있는지 확인해보았더니... 8080은 잘 되는데 5000이 안되더라...
- 무슨문제인지 이것저것 설정을 많이 바꿔보다가 혹시나 포트에 문제가 있나해서 cmd에 netstat -ano | find "5000" 을 통해 5000번 포트를 사용하고 있는게 무언가 있나 확인했다.
- 보니까 파이썬에서 연 ai서버 말고는 문제가 없다. 8080도 열어봤더니 뭔가 달랐다! 엥?? 8080은 0.0.0.0 listening하는데 5000은 127.0.0.1에서 listening 하는것이었다.
- 0.0.0.0은 모든 외부 사용자가 사용가능하도록 하는것인데 5000은 그게 안되어있으니까 외부에서 접속이 안되던것이다!
- 그래서 app.run에 추가적으로 host='0.0.0.0' 붙여줬다. 그랬더니 이제 postman에서는 http://chaeserver.iptime.org:5000/classify 을 통해서 사용이 가능하다!
- 이제 프론트엔드에서 구현하는것만 남았다... 

### 이미지 업로드
- 찾아보니 formData에 값이 없을땐 신호가 가는데 값을 넣으니 안간다! 그니까 formData가 이상한거지
- 이런저런 시도를 정말 많이했는데 결론이 났다. blob이든 fileSystem이든 뭐든 쓸 필요없다. 그냥 formData양식이 안맞아서 생긴 오류...
```
formData.append('image', {
    name: fileName,
    type: 'image/jpg',
    uri: faceImage
});
```
- 이런 양식으로 제출하니까 잘 됐다! 참고로 fileName은 faceImage (file:///data어쩌구...에서 /기준으로 split해서 pop()한것으로 실제 파일명 (23532513.jpg같은거)을 담았다.
- ![image](https://github.com/ChaeDoll/TIL/assets/108540812/6da928fb-e4cc-4226-afeb-b7efc22e8dd5)
- 깔끔하게 이런 코드가 나왔다.
- ![image](https://github.com/ChaeDoll/TIL/assets/108540812/ff1dd562-ec21-4ef9-a51f-611b394b8238)
- 이후 테스트로 내 얼굴을 찍고 결과를 봤다.
![image](https://github.com/ChaeDoll/TIL/assets/108540812/783c0d21-0b60-454e-ae8a-bef03dee1741)
![image](https://github.com/ChaeDoll/TIL/assets/108540812/3f8be2c2-58f4-4373-8a6b-b43eb7f3a993)
